pkg/k8sctrls/manager.go

```diff
@@ -3,76 +3,195 @@ package k8sctrls
 import (
 	"context"
 	"fmt"
+	"strings"
+	"sync/atomic"
+	"time"
 
+	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/client-go/kubernetes/scheme"
 	"k8s.io/client-go/rest"
+	"k8s.io/utils/pointer"
 	ctrl "sigs.k8s.io/controller-runtime"
 
 	"github.com/seal-io/walrus/pkg/dao/types"
+	"github.com/seal-io/walrus/utils/gopool"
 	"github.com/seal-io/walrus/utils/log"
 )
 
+// ManagerOptions holds the options for creating a new manager.
+type ManagerOptions struct {
+	// IsReady observes whether the manager is ready,
+	// the caller can leverage this symbol to be aware the manager's progress.
+	IsReady *atomic.Bool
+	// LeaderElection indicates whether to enable leader election.
+	LeaderElection bool
+	// LeaderLease indicates the duration of the lease that keeps the leadership.
+	LeaderLease time.Duration
+	// LeaderRenewTimeout indicates the timeout of renewing the leadership.
+	LeaderRenewTimeout time.Duration
+}
+
 func NewManager(opts ManagerOptions) (*Manager, error) {
 	logger := log.WithName("k8sctrl")
 
-	mgr, err := ctrl.NewManager(opts.K8sConfig, ctrl.Options{
-		Scheme:    scheme.Scheme,
-		Logger:    log.AsLogr(logger),
-		Namespace: types.WalrusSystemNamespace,
+	// Defaults.
+	if opts.IsReady == nil {
+		opts.IsReady = &atomic.Bool{}
+	}
+
+	if opts.LeaderLease == 0 {
+		opts.LeaderLease = 10 * time.Second
+	}
+
+	if opts.LeaderRenewTimeout == 0 {
+		opts.LeaderRenewTimeout = 5 * time.Second
+	}
+
+	// Build options for creating controller manager.
+	options := ctrl.Options{
+		// General.
+		Scheme:     scheme.Scheme,
+		SyncPeriod: pointer.Duration(time.Hour),
+		Logger:     log.AsLogr(logger),
+		Namespace:  types.WalrusSystemNamespace,
 
 		// Leader election.
-		LeaderElection:          opts.LeaderElection,
-		LeaderElectionID:        "walrus-leader-election",
-		LeaderElectionNamespace: types.WalrusSystemNamespace,
-	})
-	if err != nil {
-		return nil, fmt.Errorf("error creating kubernetes controller manager: %w", err)
+		LeaderElection:                opts.LeaderElection,
+		LeaderElectionID:              "walrus-leader-election",
+		LeaderElectionNamespace:       types.WalrusSystemNamespace,
+		LeaderElectionReleaseOnCancel: true,
+		LeaseDuration:                 pointer.Duration(opts.LeaderLease),
+		RenewDeadline:                 pointer.Duration(opts.LeaderRenewTimeout),
```

Comment:
Why we don't keep the name RenewDeadline?

Reply:
Deadline seems more like an epoch not a duration.